multicluster/controllers/multicluster/member/clusterset_controller.go

```diff
@@ -133,6 +141,14 @@ func (r *MemberClusterSetReconciler) Reconcile(ctx context.Context, req ctrl.Req
 				return ctrl.Result{}, err
 			}
 		}
+		// When the installedLeader is not empty, it means a ClusterSet was initialized at least once,
+		// we should try to clean up all stale resources before a new CommonArea is created.
+		if r.installedLeader != (leaderClusterInfo{}) {
+			klog.InfoS("Clean up all stale resources created by Antrea Multi-cluster controller", "clusterset", req.NamespacedName)
+			if err := cleanUpResourcesCreatedByMC(ctx, r.Client); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
Another part is here to ensure no stale resources are cleaned up before a new ClusterSet is created.

Reply:
The same as my above question. Does stale resource cleanup really help here?

Reply:
It has the same purpose to ensure the member cluster can be in a right status before a new common area is created. Since it won't create a common area before cleanup successfully. No new resources can be imported. It helps to avoid new data mixed with stale data in a member cluster.

Reply:
Again, do we need to do cleanup now, or we should just do cleanup at ClusterSet deletion (and let stale controller handles stale resources)?

Reply:
In K8s controller-runtime framework, the reconcile doesn't care if current request is triggered by a creation or deletion event. This is to avoid a case that when a ClusterSet is deleted but resources clean up failed at first try, meanwhile a new ClusterSet with the same ID is also created right after that, the retry caused by a failure will get a ClusterSet successfully, then it will skip deletion but start to create a CommonArea. So we'd better to clean up here as well to handle the corner case.

Reply:
This may be a valid point. Question - will stale controller take care of stale resource deletion anyway for ClusterSet recreation, even we do not clean up here?

Reply:
No, stale controller will only run when the Controller starts and run once if there is no error. It will retry only when there is any error. So we rely on ClusterSet deletion event to trigger resource clean up in ClusterSet controller when there is any ClusterSet recreation.

Reply:
Do you think it makes sense to run stale controller here the same as the MC Controller startup case?

Another question for ResourceExports - I thought we have a mechanism to delete stale ResourceExports for the member re-join case, but is that true? How that is implemented?

Reply:
Stale controller in member cluster will check leader cluster data and remove any local stale data correspondingly. It won't clean up local resources if there is any corresponding ResourceImports in the leader. Here we want to clean up all old imported resources before a CommonArea is created. They have different purpose.

When you say 're-join case', do you mean delete a ClusterSet and create a ClusterSet CR later in a member cluster?
For now, the member_cleanup_controller.go will be responsible to clean up all stale ResourceExports when the MemberClusterAnnounce CR is deleted.
MemberClusterAnnounce CR deletion will be called the member cluster controller when a ClusterSet CR is deleted, or it will be deleted by the stale controller in leader if there is no timestamp update over 24 hours.

Reply:
For imported resources, how if MC Controller restarts before cleanup succeeds?
 
For ResourceExports, how we handle the case MemberClusterAnnounce is recreated before ResourceExports cleanup succeeds? And how about MC Controller restarts before ResourceExports cleanup completes?

Reply:
Yeah, I think we need to handle the cleanup again if the MC controller restart. I will refine the code to clean up all resources in the stale controller if there is no ClusterSet CR at all.

if MemberClusterAnnounce is recreated before ResourceExports cleanup succeeds, then there will be stale ResourceExports, this is a valid case, I will check how to clean up resources in this case. Thanks.

if MC Controller restarts before ResourceExports cleanup completes, there will be an one time clean up `RunOnce` in `member_cleanup_controller` to cleanup all stale ResourceExports if there is no corresponding MemberClusterAnnounce.

Reply:
For imported resources, should we consider using stale controller to do cleanup in all cases, if stale controller anyway needs to do that after restart?

The same for ResourceExports, we may think about a unified way to do cleanup in all cases.

Reply:
If we use current stale controller to cover all cases, it means we need to set up a sync mechanism to stop the clean up process in the stale controller when there is a common area is initialized by the ClusterSet controller. I feel it may make things complicated. I refined a little bit to make ClusterSet controller to call Cleanup method on different cases.
stale controller will be responsible for restart case only.

For the case `if MemberClusterAnnounce is recreated before ResourceExports cleanup succeeds, then there will be stale ResourceExports`, I think there is no way to know what are stale ResourceExports in a leader cluster if there is corresponding MemberClusterAnnounce. The member cluster should be the single source of truth. So I added a step here https://github.com/antrea-io/antrea/pull/5351/files#diff-d2f1e28c34f19f40190c020679752f63a69522cc837119c24417ef10e6b528feR371-R380 to do a clean up before we set a ClusterSet as ready.
The member cluster controllers won't create any ResourceExports before the ClusterSet is ready when the ClusterSet is rejoining in this PR https://github.com/antrea-io/antrea/pull/5410

Reply:
> If we use current stale controller to cover all cases, it means we need to set up a sync mechanism to stop the clean up process in the stale controller when there is a common area is initialized by the ClusterSet controller. I feel it may make things complicated. I refined a little bit to make ClusterSet controller to call clean up methods in stale controller for different cases. stale controller will be responsible for restart case only.

Your current code is different what you described above. In your current code, stale controller indeed cleans up all import resources if no ClusterSet. In this case, there is no need to do cleanup in ClsuterSet controller again. For synchronization between controllers, I think there are other cases too (could you confirm this), so it is a problem we anyway need to solve?

> For the case if MemberClusterAnnounce is recreated before ResourceExports cleanup succeeds, then there will be stale ResourceExports, I think there is no way to know what are stale ResourceExports in a leader cluster if there is corresponding MemberClusterAnnounce. The member cluster should be the single source of truth. So I added a step here https://github.com/antrea-io/antrea/pull/5351/files#diff-d2f1e28c34f19f40190c020679752f63a69522cc837119c24417ef10e6b528feR371-R380 to do a clean up before we set a ClusterSet as ready.

I feel we should not do cleanup when updating ClusterSet status, which is hard to understand. Why we cannot wait for the cleanup to be done by ClusterSet controller?